### PR TITLE
Implementation

### PR DESCRIPTION
**Pull request by neachdainn**
This is not thoroughly tested, but it appears to work at least a little. Merge whenever.

### CODE REVIEW

File: lcm/src/lcm/spsc.rs [hunk 1]
```diff
@@ -0,0 +1,459 @@
...
+        done_p.send(()).unwrap();
+    }
+
+    #[test]
+    fn padding() {
+        // Before Rust 1.24, using `#[repr(C, packed)]` did not require blocks
+        // of unsafe code. Now that it does, trying to add those blocks would
+        // be both a pain and ugly, particularly since I am very confident that
+        // I manually padded things correctly. This test confirms that there is
+        // no padding in the struct so the absense of `packed` changes nothing.
```
Comment (neachdainn) [inline lcm/src/lcm/spsc.rs]: Spelling error. "Absense"

**Comment by adeschamps** [comment]
Nice! I'll have to spend some more time looking at this over the weekend.

It looks like it's hardcoded for UDP. The C implementation of LCM supports five different types of providers:
- file://
- memq://
- mpudpm://
- tcpq://
- udpm://

I just use udpm and file, so personally those are the ones I'm interested in implementing first. If we can separate this into a trait and a struct, sort of like this, then we can support multiple backends:

```rust
pub trait Provider {
    // stuff...
}

pub struct Lcm {
    provider: Box<Provider>,
}
```

I already wrote some code to handle the construction of various providers, so that they can be enabled or disabled with Cargo feature flags. I'll make a PR to your fork (probably tomorrow night) that puts some of that infrastructure in place. After that, I actually won't mind merging unfinished work as long as it's behind a feature flag that's disabled by default.

In the meantime, feel free to continue what you're doing, since my changes shouldn't affect it in any significant way.

**Comment by neachdainn** [comment]
I don't think we need to support all the providers. Even though there is code for them in the LCM repo, the only ones that [the docs](https://lcm-proj.github.io/group__LcmC__lcm__t.html#gaf29963ef43edadf45296d5ad82c18d4b) say are supported are `udpm`, `file`, and `memq`. Since `memq` is really only useful for testing, I don't think we need to support that.

Regardless, it should be easy to set up the structure for implementing different providers. The way I have the code laid out right now, we might not even need to set up a `Provider` trait. I currently have the receiving end of the provider communicating the the `Lcm` object through a pair of MPSC channels.

If we have the outbound messages also use MPSC channels, the `Lcm` object is totally agnostic to how the messages are being handled (and we will become significantly more ROS-like). We could handle that through a separate sending thread (extra thread, simpler code) or by depending on something like Mio (only one background thread, significantly more complicated code).

Alternatively, we could either set things up as `Lcm<P: Provider>` or use modules (e.g., `lcm::udpm::Lcm` to avoid having a trait object.

**Comment by neachdainn** [comment]
The more I think about it, the more I like the idea of:

```rust
enum Provider {
    Udpm,
    File,
    ..
}

impl Lcm {
    fn with_provider(p: Provider) -> Self {
        ..
    }
}
```

At that point, we can either make `Lcm` totally provider agnostic or just implement all of the provider methods as part of the enum.

**Update:** I am beginning work on the enum version. I'm open to discussion, I just happen to have a lot more free time this week than normal and I'm antsy to get coding.

**Comment by neachdainn** [comment]
Based on the `lcm-gen` crate, I'm assuming we want to use Failure for the `lcm-rs` crate too. There are a few places where the code should return an error but an `io::Result` doesn't seem quite appropriate. In particular, the `UdpmProvider::subscribe` should probably indicate a failure to subscribe if the `Backend` has failed.

**Comment by adeschamps** [comment]
I'm not too fond of the VTable enum. It doesn't allow us to express anything that can't be achieved by a trait, and it will result in the code for the various providers being interleaved rather than in their own files. It's essentially just a manual implementation of dynamic dispatch.

Another thing, although I may be wrong about this, but I don't think that the subscribe/unsubscribe logic needs to extend all the way to the individual providers, since it'll probably be identical regardless of the backend. I suspect that the providers will just have to be able to send/receive something like this:

```rust
pub struct Buffer {
    pub time: u64,
    pub channel: String,
    pub data: Vec<u8>,
}
```

A lot of the pub/sub, encoding and decoding can probably be implemented by methods on the root Lcm struct.

And yes, you're right about error handling. The original implementation used io::Error for everything. Sometimes it was appropriate, and sometimes not. It was written before error-chain was popular, and before failure was around at all.

**Comment by neachdainn** [comment]
> It doesn't allow us to express anything that can't be achieved by a trait

What advantages does a trait give us over the enum? If this were something we were exposing to the public API a trait would definitely be the way to go. However, we only have a finite, known set of providers to implement. This seems like the perfect case for the enum, plus it has the added benefit of not requiring `Lcm<P: Provider>` or the use of a `Box<Provider>`.

> It's essentially just a manual implementation of dynamic dispatch.

Yes and no. Yes, because it does kind of act like that and is a lot of boilerplate. No, because this is specifically to avoid the dynamic dispatch of a trait object.

> I suspect that the providers will just have to be able to send/receive something like this:

That would work but it forces us to decide between (a) allocating a new `Vec<u8>` and a new `String` every time we have an incoming message, (b) making the code significantly more complicated, or (c) using a lock. Decoding the message in the provider allows us to use some form of array based queue.

> And yes, you're right about error handling.

I am knee-deep in converting the `Message`/`Marshall` traits and `lcm-derive` to use Failure. I'll push something before too long.

**EDIT:** The `VTable` is also largely based on how the C version of LCM does things. As far as I can tell, the `lcm_t` does not do much more than lock a few locks and then call into the `lcm_provider_vtable_t`.

**EDIT:** Using the new error type is causing the `quote` trait to hit the recursion limit. I'm not sure why, but I'll push the commit regardless if I can't figure it out soon.

**Comment by adeschamps** [comment]
Aside from allowing the admittedly unlikely use case of a user providing their own implementation, I think traits give us two nice things.

One is a separation of code for the different implementations.

The other is that they actually allow us to avoid dynamic dispatch even more than the enum approach does - traits can be used for both runtime and compile time polymorphism.

We could define
```rust
struct Lcm<P: Provider = Box<Provider>> { provider: P }
```
 and then
```rust
impl<P:Provider> Provider for Box<P> { /* ... */ }
```

Although in the default case, we would read the `LCM_DEFAULT_URL` environment variable and construct a `Box<Provider>`, it would also allow you to explicitly instantiate `Lcm<UdpmProvider>` if you know that's what you want, in which case you can avoid both dynamic dispatch and matching on enums entirely.

**Comment by adeschamps** [comment]
Regarding allocations, I might still be wrong, but it's possible that the providers could return a reference to `Buffer`, and keep using the same memory to receive the next messages.

Whether decoding messages requires allocation each time is a separate question - although I lean towards giving the subscriber a reference to each message, and letting them clone it if needed.

**Comment by neachdainn** [comment]
> One is a separation of code for the different implementations.

I'm not sure I 100% understand what you mean. The way it's currently set up does allow for the separation of code. The `VTable` object should not do more than what it is already doing.

> traits can be used for both runtime and compile time polymorphism

There is one major downside of using traits: they're infectious. If I have a struct that I want to take an `Lcm` object then I either have to commit to a specific provider or that struct suddenly becomes generic itself (or a `Box<Lcm>` + and `Lcm` trait). Same goes for functions. With the current enum version, using an `Lcm` in the struct is much less work. 

Honestly, now that I'm thinking about it, this may trump the dynamic dispatch issue for me. Trying to use a non-provider-specific `Lcm<P>` object really could be a hassle.

> Although in the default case, we would read the `LCM_DEFAULT_URL` environment variable and construct a `Box<Provider>`

I would like there to be some "default" function that returns an `Lcm` instance with the default provider (no DD, which means no `LCM_DEFAULT_URL`) in addition to one that takes an `lcm_url` and falls back onto the environment variable. Basically 100% of my LCM usage is with the default settings on the default provider and I would rather not have to worry about setting that up every time.

> Regarding allocations, I might still be wrong, but it's possible that the providers could return a reference to `Buffer`, and keep using the same memory to receive the next messages.

I don't think there's an easy way to do that without locks. There's definitely a complicated way, though. It would likely involve swapping buffers into and out of the receiving struct. That also effectively turns the channels into linked lists (with all the downsides that come along with) versus a contiguous array.

---

Really, I don't think anything that's been proposed here is bad. I just think that the ease of the enum vastly outweighs the cost of a `match` statement while the advantages of the provider generic are outweighed by the infectious nature of generics and the dynamic dispatch that would come along with a default provider.

---

**EDIT:** ~~The way I have this code planned out keeps the code for subscribe/unsubscribe/handle logic in the same file as the `Lcm` struct. The struct is going to construct closures to send to the backend.~~ Nope. I am tired and was wrong. But the code I was trying to talk about is now checked in.

Which is kind of funny, considering how much we've talked about dynamic dispatch in this thread.

**Comment by neachdainn** [comment]
One last comment before I call it a night: do you know if the file provider would use a separate thread for receiving?

I haven't looked into the C version at all and I've never used it, so I don't know what it does. But I imagine that it would work fine without a reader thread, so using channels to communicate in the provider might be overkill. Which could be an argument for keeping code out of the `Lcm` object and in the `Provider` (trait or enum).

**Comment by adeschamps** [comment]
I'd like to actively discourage any use of LCM that *doesn't* read the `LCM_DEFAULT_URL` variable. I've been handed binaries in the past that had hardcoded URLs. It only takes one to be an inconvenience to an entire system. (fun fact - sometimes it's possible to modify the binary with hexedit)

Of course, you should still be able to explicitly pick a provider - I just don't want that option displayed as prominently in the docs.

**Comment by adeschamps** [comment]
I'm coming around to the VTable enum approach. I agree that the "match on enum" implementation of dynamic dispatch would probably be faster than using a boxed trait object.

That said, I think that the whole `impl VTable` block is an unnecessary layer of indirection (in terms of code - the compiler will almost certainly inline it). Unless you have something else planned, I would move the definition of the VTable enum into same file as the Lcm struct.

Let's not bother with this right now, but later we can add some attributes to conditionally compile the different VTable variants:
```rust
enum VTable {
    #[cfg(feature = "file")]
    File(FileProvider),
    #[cfg(feature = "udpm")]
    Udpm(UdpmProvider),
    // ...
}
```
Those features would be enabled by default, but if you wanted to disable all but one then you end up with a single-variant enum, which will eliminate all dynamic dispatch without requiring the generic parameter that I proposed earlier.

**Comment by neachdainn** [comment]
> I've been handed binaries in the past that had hardcoded URLs. It only takes one to be an inconvenience to an entire system.

That's a very good point. I'm so used to being the person who writes all of the code that I didn't think of that. I'll add the parsing of the `LCM_DEFAULT_URL` first thing.

>  Unless you have something else planned, I would move the definition of the VTable enum into same file as the Lcm struct.

I'll get this done second thing.

---

I'll have to go back to my normal work schedule on Sunday, so I'm going to try and get through as much as this code as possible today and tomorrow.

**Comment by neachdainn** [comment]
~~The `lcm::Provider` enum is now gone. I couldn't come up with a good way to allow the network portion for the UDPM be a hostname without a bad compromise. I'll push it as soon as it builds, but the `Lcm` object now has a `with_provider(lcm_url: &str)` method.~~

After looking through the code and playing around some, it seems like the UDPM can only handle an LCM URL that is an IPv4 address. It can't handle IPv6 or names, so I'm not going to try and support those at this time.

File: lcm/src/lcm/mod.rs [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+use regex::Regex;
+
+mod providers;
+use self::providers::udpm::UdpmProvider;
+
+use Message;
+use error::*;
+
+/// Convenience macro for dispatching functions among providers.
+macro_rules! provider
```
Comment (adeschamps) [inline lcm/src/lcm/mod.rs:13]: I really like this macro.

Comment (neachdainn) [reply lcm/src/lcm/mod.rs:13]: I was really pleased to find that it was a valid macro.

**Comment by neachdainn** [comment]
How do you envision the File provider working? Specifically, will it use a background thread to monitor the file or will that happen from the main thread? If it's happening in the main thread then I want to keep a lot of the logic in the providers so that the File provider doesn't pay unnecessary overhead. If it uses a separate thread, I can move thing from the `UdpmProvider` and into the `Lcm`.

**Comment by adeschamps** [comment]
Let's go with the simpler implementation and assume it'll use a separate thread so that more of the logic can go in the `Lcm` impl. Playback from a log file is unlikely to be a performance bottleneck.

**Comment by neachdainn** [comment]
I think we're going to have to implement some form of interior mutability. Writing an echo server is far more complicated than it should be without it.

I.e., the following does not compile.
```rust
let mut lcm = Lcm::new()?;
lcm.subscribe("channel", |m| lcm.publish("channel_reply", m))?;
```

We also probably want to consider a `subscribe_bytes` function that allows for a callback on the raw bytes of the message. It should have a better name, though.

**Comment by neachdainn** [comment]
I have done almost no testing on this but I am able to send messages between processes. Merge whenever and I'll open up issues for the things that I think need to be fixed/changed.

**Comment by adeschamps** [comment]
So, I've had a busy weekend and have somewhat lost track of this.

It looks like you're right about interior mutability. Conceptually, publishing doesn't mutate the state of the `Lcm` object from the perspective of the user. As far as I can tell, it's only the `sequence_number` field that needs interior mutability.

**Comment by adeschamps** [comment]
I've been looking forward to contributing more, but you've been getting so much done over the last few days, I could hardly keep up.

Merging now!

**Comment by neachdainn** [comment]
My vague plan was that I would leave the File provider to you since I really don't use it. Between that and the fact this the UDPM provider is a bit messy, there's still plenty left to do!

**Pull request merged**